### PR TITLE
[invoke_subgraph] Fix get_output_metadata requires_grad bug

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #178532

`meta["val"]` is always populated via `snapshot_fake → detach()`, which
strips `requires_grad`. This caused `get_output_metadata` to incorrectly
mark all float outputs as no-grad (0 tangents in backward) when taking
the static metadata path (e.g. in `reenter_make_fx`).

For float/complex tensor outputs, fall back to
`_get_output_metadata_by_execution` which checks `requires_grad` on the
actual executed output. The fallback check is hoisted into a single
pre-scan before the main loop to avoid redundant subgraph executions.

Authored with Claude.

### CODE REVIEW

File: torch/_higher_order_ops/invoke_subgraph.py [hunk 1]
```diff
@@ -443,35 +443,52 @@ def get_output_metadata(subgraph, *operands):
...
-
         val = output_arg.meta["val"]
         if isinstance(val, torch.SymInt):
             output_metadata.indexes_with_symint.add(idx)
             output_metadata.indexes_with_no_grad.add(idx)
         elif isinstance(val, torch.Tensor):
-            # Check if tensor requires grad from metadata
-            if hasattr(val, "requires_grad") and not val.requires_grad:
-                output_metadata.indexes_with_no_grad.add(idx)
+            if val.requires_grad:
```
Comment (ydwu4) [inline torch/_higher_order_ops/invoke_subgraph.py:487]: hm.. i'm kind of confused, we're returning runtime error if out tensor requires_grad?

Comment (tugsbayasgalan) [reply torch/_higher_order_ops/invoke_subgraph.py:487]: There is no way this can be triggered in general because int tensors don't have grad_fn. This looks weird because we can't do assert xyz pattern anymore.

Comment (ydwu4) [reply torch/_higher_order_ops/invoke_subgraph.py:487]: maybe just changed to tensor dtype is int? If any floating tensor output, we would do an execution

**Review by ydwu4 — APPROVED** [review]

File: torch/_higher_order_ops/invoke_subgraph.py (hunk 1 above)
Comment (tugsbayasgalan) [reply torch/_higher_order_ops/invoke_subgraph.py:487]: we check the dtype before this loop and execute the op manually.

**Review by mlazos — APPROVED** [review]
